api/v1alpha4/machinedeployment_types.go

```diff
@@ -35,6 +35,10 @@ const (
 	// RevisionAnnotation is the revision annotation of a machine deployment's machine sets which records its rollout sequence.
 	RevisionAnnotation = "machinedeployment.clusters.x-k8s.io/revision"
 
+	// RolledoutAfterAnnotation annotates the MachineTemplateSpec to trigger a
+	// MachineDeployment rollout when the RolloutAfter criteria is met.
+	RolledoutAfterAnnotation = "machinedeployment.clusters.x-k8s.io/RolledoutAfter"
```

Comment:
Instead of relying on a label, could we add additional logic like we have in KCP to take into account the `Spec.RolloutAfter` field?

Reply:
`MachineDeployment.Spec.RolloutAfter` is the source of truth and the annotation is an impl detail https://github.com/kubernetes-sigs/cluster-api/pull/4596/files#diff-34f64847a48c9fc1b3222c9bec6c93f30767a93262215eb4305d78c3b9097c63R225.

Orchestrate this "manually" as in KCP does not seems optimal here since Machines are owned by MachineSets and subject to its reconciliation logic and to MachineDeployment replicas so leveraging the MachineDeployment might be the most viable path.

Not sure if you were suggesting something different?

Reply:
The rollout is determined in this function https://github.com/kubernetes-sigs/cluster-api/blob/master/controllers/mdutil/util.go#L416-L431

Today we're only looking at the MachineTemplate, we could add some logic there to compare `CreationTimestamp` with `RolloutAfter`?

Reply:
I though about that but I couldn't think of a working way in that direction so far because `mdutil.FindNewMachineSet(d, msList)` prefers the oldest MachineSet in case of equal templates., see my reasoning here https://github.com/kubernetes-sigs/cluster-api/pull/4596#discussion_r705525941

Reply:
> The issue with the above is that mdutil.FindNewMachineSet(d, msList) seems to prefer oldest MachineSet in case of equal templates.

From the text in there, the assumption made is that usually there should be only one MachineSet with the same template, otherwise we wouldn't need to roll out change; there is an edge case where there could be more than one MachineSet with the same template, and the code always picks the oldest one (which might already be scaled).

The above shouldn't impact adding new logic that looks at the creation timestamp and compares with a `rolloutAfter` field?

Reply:
The workflow as I can think of would be:
- `if spec.rolloutAfter.before(now)` then we create a new machineSet.
- The new Machine would have the same template than the current one.
- The next iteration `FindNewMachineSet` would wrongly pick the old MachineSet rather than the one we just created and that need to be scaled up.

Reply:
What if `FindNewMachineSet` also takes into account the `rolloutAfter` field and discards MachineSets which are too old?

Reply:
I guess we could let `FindNewMachineSet` to also find MachineSets by name (which contains rolloutAfter info hashed) and prefer that one in case of template equality. That would mostly defeat the ordering by oldest for the edge case described in the func comment though, e.g once rollouAfter took place at some point (so it's permanently hashed in the name) when that edge case happens `FindNewMachineSet` will always prefer the new MachineSet since that one would be name match.

Everything I can think of in this line feels a little hacky, I'm not sure what would be the value of trying to modify existing logic vs just leveraging it and build atop with the annotation.

Reply:
> The next iteration FindNewMachineSet would wrongly pick the old MachineSet rather than the one we just created and that need to be scaled up.

Wouldn't we also check the CreationTimestamp now at every iteration though? If there is a RolloutAfter, the MachineSet CreationTimestamp must be after the RolloutAfter's value. Even if templates are the same, we shouldn't pick the oldest one in that case.

We should make sure to cover edge cases 

> Everything I can think of in this line feels a little hacky, I'm not sure what would be the value of trying to modify existing logic vs just leveraging it and build atop with the annotation.

Adding an annotation or label to a MachineDeployment to rollout a template is truthfully something that might change in the future. I'd rather find a better way to rollout changes without using any behaviors, but rather be explicit about it when we need to calculate changes. In the future we might want to add more logic to determine if a MachineSet needs to be deployed, and relying on annotations is probably not going to work out long term.

Reply:
>Wouldn't we also check the CreationTimestamp now at every iteration though? If there is a RolloutAfter, the MachineSet CreationTimestamp must be after the RolloutAfter's value. Even if templates are the same, we shouldn't pick the oldest one in that case.

Yes we could do that but then we'll be changing `FindNewMachineSet` to always pick the newest even in the [duplicated template edge case](https://github.com/kubernetes-sigs/cluster-api/blob/master/controllers/mdutil/util.go#L421-L424). We have no ability to differentiate between the duplicated template edge use case and the rolloutAfter use case, so for both we'll be picking the newest one. If we don't care about the edge use case, then it might work. I tried to summarise that reasoning here https://github.com/kubernetes-sigs/cluster-api/pull/4596#discussion_r706022157.

Reply:
> Yes we could do that but then we'll be changing FindNewMachineSet to always pick the newest even in the duplicated template edge case.

The duplicate one is truthfully an edge case that should be audited, but that's probably something for a different issue/PR. The code is [the same as the ReplicaSet code upstream](https://github.com/kubernetes/kubernetes/blob/cf535b03390608b90d1a03701c304319bc1f6d61/staging/src/k8s.io/kubectl/pkg/util/deployment/deployment.go#L183-L197), which probably has this issue because api-server and controller-manager could be restarting and duplicating ReplicaSet(s). I don't know if this can happen for MachineSets.

> We have no means to differentiate between the duplicated template edge use case and the rolloutAfter use case, so for both we'll be picking the newest one.

This shouldn't apply in case of RolloutAfter. If a rollout time is in effect, even if we have duplicates, we've already rolled out a new MachineSet at that point. 

If there are duplicates (depending on whatever RevisionHistory is set to), `cleanupDeployment` function takes care of deleting outdated MachineSets.

Reply:
The edge case issue was because different kube versions resulted  in different hashes for ReplicaSet names resulting in duplicated replicasets.

>To further clarify though, we wouldn't be changing the code to pick the newest one, but the oldest one with these conditions:
ms.metadata.creationTimestamp > md.spec.rolloutAfter
EqualMachineTemplate(md, ms) == true

Ok this makes sense let me implement it and keep discussing through code.

Reply:
@vincepri @sbueringer updated PTAL.

Reply:
Updated to test and support the following possible cases:
- RolloutAfter == nil
  - Pick oldest
- RolloutAfter < Now
  - Oldest 
  - Created > last RolloutAfter
- RolloutAfter > Now
  - Oldest 
  - Created > last RolloutAfter
  
To keep track of the last rolloutAfter I'm using an annotation `LastRolloutAfterAnnotation`. We could use a status field but I'd rather wait for the implementation to get mature and see how/if it combines with things like maxAge.

@vincepri @sbueringer PTAL.